### ISSUE TITLE
`select for update skip locked` for edgedb backed queues

### ISSUE THREAD

**Issue by MarceColl** [edgedb/edgedb#5316]
Hello!

EdgeQL seems to be an amazing productivity tool, and the whole stack is fantastic so thank you for that! But I haven't found a way to represent queues in the system.

I have some tables that are basically queues, this allows me to enqueue stuff in an atomic way. A common pattern for this type of operations is to do a `SELECT FOR UPDATE SKIP LOCKED` in postgres in the consumers, this allows concurrent consumers to get items from the queue without locking. I've looked around the EdgeQL spec & docs and in the issues here but I haven't found anything. Is there any currently existing feature that would allow me to do this? In case there is not, is there anything planned?

Thank you!

**Comment by elprans** [comment]
No support for explicit locks yet.  It is, indeed, a very useful feature, we'll look into the feasibility of adding it to EdgeQL in 4.0.

**Comment by dvv** [comment]
Wonder if there's an update for this?

**Comment by AlexDaniel** [comment]
@elprans as 4.0 is now released, what's the new plan for the locks?

**Comment by scotttrinh** [comment]
> Hitting this. Sometimes we have a C in app calculation & row insert that depends on row A & B. There's no way currently to row-lock/mark row A & B with for update. Serializable transactions do not solve this.

Can you share something a bit more concrete here? Serializable should put predicate locks on any reads within the query, so it's possible that just adding some selects and assertions in the query will cause A and B to be "locked" (or at least updates serialized).

**Comment by CarsonF** [comment]
@scotttrinh Ok I'll confess I haven't actually tried, just trying to create the plan. We have a `readWithLock()` that I need to implement with EdgeQL.
Are you saying PG understand to lock this user if I `select <User><uuid>"..."`, and later in transaction `insert Foo {}`?
Or an assertion is needed? like how?

Also I'm not entirely sure "serialized" is sufficient. I would need the 2nd parallel transaction to throw/retry the read on that user, so that the app can recalculate the Foo insert. (All inside `client.transaction(...)`) Is that how it works?

**Comment by scotttrinh** [comment]
Yeah, it's a little bit tricky to explain simply (due to my lack of mastery here), but [how Postgres' `SERIALIZABLE` works is via predicate locking (implemented using "Serializable Snapshot Isolation" to minimize locks)](https://www.postgresql.org/docs/current/transaction-iso.html#XACT-SERIALIZABLE):

> To guarantee true serializability PostgreSQL uses predicate locking, which means that it keeps locks which allow it to determine when a write would have had an impact on the result of a previous read from a concurrent transaction, had it run first. In PostgreSQL these locks do not cause any blocking and therefore can not play any part in causing a deadlock. They are used to identify and flag dependencies among concurrent Serializable transactions which in certain combinations can lead to serialization anomalies.[...]
> Predicate locks in PostgreSQL, like in most other database systems, are based on data actually accessed by a transaction.

So, I think you can get consistent locking behavior if you query for invariants that need to hold true between different concurrent transactions. It's _possible_ that it's not something you can write in your query, and it's not as straightforward as taking an explicit lock, but I believe throughput is increased substantially if there is much contention. Without a concrete case, it's hard to be more ... concrete, but I feel like there is probably some good literature out there on this. The best I've found is the chapter on Transactions in [Designing Data-Intensive Applications](https://dataintensive.net/).

Some random advice from that book that might apply for your case:

> Contention can be reduced with commutative atomic operations, for example, if several transactions concurrently want to increment a countier, it doesn't matter in which order the increments are applied (as long as the counter isn't read in the same transaction), so the concurrent increments can all be applied without conflicting.

---

But, yeah, I don't think there is a mechanical equivalent to `readWithLock()` other than opening a transaction. It might be close to open a transaction and read the full data, but you need to try doing a write in that same transaction to trigger the serializable mechanism to cause the concurrent queries to retry. No real generic solution there until we add explicit locking.
